L1Trigger/L1THGCal/plugins/be_algorithms/HGCalSimCluster.cc

```diff
@@ -0,0 +1,256 @@
+#ifdef DEBUG
+                cout<<"[HGCalTriggerSimCluster]::["<<__FUNCTION__<<"] DONE"<<endl;
+#endif
+            }
+
+            // run, actual algorithm
+            virtual void run( const l1t::HGCFETriggerDigiCollection & coll,
+                    const edm::ESHandle<HGCalTriggerGeometryBase>&geom,
+		            const edm::Event&evt
+                    )
+            {
+#ifdef DEBUG
+                cout<<"[HGCalTriggerSimCluster]::[run] Start"<<endl;
+#endif
+                //1. construct a cluster container that hosts the cluster per truth-particle
+                std::unordered_map<uint64_t,std::pair<int,l1t::HGCalCluster> > cluster_container;// PID-> bx,cluster
+                evt.getByToken(sim_token,sim_handle);
+
+                if (not sim_handle.isValid()) { std::cout<<"[HGCalTriggerSimCluster]::[run]::[ERROR] PFCluster collection for HGC sim clustering not available"<<std::endl; throw 39;}
+
+                // 1.5. pre-process the sim cluster to have easy accessible information
+#ifdef DEBUG
+                cout<<"[HGCalTriggerSimCluster]::[run] processing sim clusters"<<endl;
+#endif
+                // I want a map cell-> [ (pid, fraction),  ... 
+                std::unordered_map<uint32_t, std::vector<std::pair< uint64_t, float > > > simclusters;
+                for (auto& cluster : *sim_handle)
+                {
+                    auto pid= cluster.particleId(); // not pdgId
+                    const auto& hf = cluster.hits_and_fractions();
+                    for (const auto & p : hf ) 
+                    {
+                        simclusters[p.first].push_back( std::pair<uint64_t, float>( pid,p.second) ) ;
+                    }
+                }
+                
+#ifdef DEBUG
+                cout<<"[HGCalTriggerSimCluster]::[run] Run on digis"<<endl;
+#endif
+
+                //2. run on the digits,
+                for( const auto& digi : coll ) 
+                {
+                    DATA data;
+                    digi.decode(codec_,data);
+                    //2.A get the trigger-cell information energy/id
+                    const HGCTriggerDetId& moduleId = digi.getDetId<HGCTriggerDetId>(); // this is a module Det Id
+
+                    // there is a loss of generality here, due to the restriction imposed by the data formats
+                    // it will work if inside a module there is a data.payload with an ordered list of all the energies
+                    // one may think to add on top of it a wrapper if this stop to be the case for some of the data classes
+                    for(const auto& triggercell : data.payload)
+                    { 
+                            if(triggercell.hwPt()<=0) continue;
+
+                            const HGCalDetId tcellId(triggercell.detId());
+                            //uint32_t digiEnergy = data.payload; i
+                            auto digiEnergy=triggercell.p4().E();  
```

Comment:
`triggercell.energy()` gives the energy of the 4-vector (so same as `triggercell.p4().E()`)